exercises/Daniel_Murphy/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,264 @@
+{
+  double a = 0.5 * b * h;
+  return a;
+}
+
+double Triangle::get_perimeter()
+{
+  double p = b + 2*sqrt(pow(h,2) + (pow(b,2)/4));
+  return p;
+}
+
+std::string Square::get_type()
+{
+  return type;
+}
+
+int Square::get_sides()
+{
+  return sides;
+}
+
+double Square::get_area()
+{
+  return pow(s1,2);
+}
+
+double Square::get_perimeter()
+{
+  return s1*4;
+}
+
+
+std::string Rectangle::get_type()
+{
+  return type;
+}
+
+int Rectangle::get_sides()
+{
+  return sides;
+}
+double Rectangle::get_perimeter()
+{
+  double p = (2*s1) + (2*s2);
+  return p;
+}
+double Rectangle::get_area()
+{
+  double a = s1 * s2;
+  return a;
+}
+
+
+std::string Circle::get_type()
+{
+  return type;
+}
+
+int Circle::get_sides()
+{
+  return sides;
+}
+double Circle::get_perimeter()
+{
+  double p = 2 * M_PI * r;
+  return p;
+}
+double Circle::get_area()
+{
+  double a = 4 * M_PI * pow(r,2);
+  return a;
+}
+
+
+void ShapeSorter::add_new(Triangle a)
+{
+  shapes.emplace_back(std::make_unique<Triangle>(a.b,a.h));
+}
+void ShapeSorter::add_new(Square a)
+{
+  shapes.emplace_back(std::make_unique<Square>(a.s1));
+}
+void ShapeSorter::add_new(Rectangle a)
+{
+  shapes.emplace_back(std::make_unique<Rectangle>(a.s1,a.s2));
+}
+void ShapeSorter::add_new(Circle a)
+{
+  shapes.emplace_back(std::make_unique<Circle>(a.r));
+}
+
+void ShapeSorter::all_of_one_type(std::string type_requested)
+{
+  std::vector<int> type_vec;
+  int number = 1;
+  for (auto const& el : shapes)
+  {
+    std::string shape_type = el->get_type();
+    if (shape_type == type_requested)
+    {
+      type_vec.push_back(number);
+    }
+    number++;
+  }
+  if (type_vec.size() == 0)
+  {
+    std::cout << "None of shape type " << type_requested << " were found.";
+  }
+  else
+  {
+    std::cout << "The shapes of type " << type_requested << " are:  ";
+    for (auto const& el: type_vec)
+    {
+      std::cout << el << "  ";
+    }
+  }
+  std::cout << "\n";
+}
+
+void ShapeSorter::match_number_of_sides(int sides_requested)
+{
+  std::string plural = "sides";
+  if (sides_requested == 1)
+  {
+    plural = "side";
+  }
+  std::vector<int> side_vec;
+  int number = 1;
+  for (auto const& el: shapes)
+  {
+    int num_sides = el->get_sides();
+    if (num_sides == sides_requested)
+    {
+      side_vec.push_back(number);
+    }
+    number++;
+  }
+  if (side_vec.size() == 0)
+  {
+    std::cout << "No shapes with " << sides_requested << " sides were found.";
+  }
+  else
+  {
+    std::cout << "The shapes with " << sides_requested << " " << plural << " are:  ";
+    for (auto const& el: side_vec)
+    {
+      std::cout << el << "  ";
+    }
+  }
+  std::cout << "\n";
+}
+
+void ShapeSorter::area_descending()
+{
+  std::cout << "\nShape Number      Area / sq u \n";
+  std::cout << "############      ###########\n\n";
+
+  std::map<double, int> area_map;
+  int number = 1;
+  for (auto const&  el : shapes)
+  {
+    double shape_area = el->get_area();
+    area_map[shape_area] = number;
```

Comment:
The link between the shapes and the numbers you're using as an identifier feels a little bit loose. It would be nicer if you had a method on the shape called getID() or similar that returned the number. I think that would probably mean you'd have to supply the ID or make the ShapeSorter generate it when the shape is added to the ShapeSorter